File: TableViewReview/TableViewReview/CompinentTableView/OCComponent/OCComponentTableView/BaseAssemblyDispatcher.m
//
//  BaseAssemblyDispatcher.m
//  TableViewReview
//
//  Created by Jialin Chen on 2019/7/31.
//  Copyright © 2019年 CJL. All rights reserved.
//

#import "BaseAssemblyDispatcher.h"
#import "BaseAssemblySectionModel.h"
#import "BaseAssemblyComponent+Private.h"

@interface BaseAssemblyDispatcher ()

@property (strong, nonatomic) NSMutableDictionary *assemblyComponents;

@end

@implementation BaseAssemblyDispatcher
@synthesize tableView, dataDictionary;

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self initWithData];
    }
    return self;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        [self initWithData];
    }
    return self;
}

- (void)initWithData {
    self.assemblyComponents = [NSMutableDictionary dictionary];
}

- (void)loadView {
    CGRect rect = [UIScreen mainScreen].bounds;
    self.view = [[UIView alloc] initWithFrame:CGRectMake(0, 0, rect.size.width, rect.size.height)];
    self.view.backgroundColor = [UIColor whiteColor];
    self.view.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    
    self.tableView = [[BaseAssemblyTableView alloc] initWithFrame:self.view.bounds style:UITableViewStylePlain];
    self.tableView.backgroundColor = [UIColor clearColor];
    self.tableView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    self.tableView.baseDelegate = self;
    [self.view addSubview:self.tableView];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self initWithData];
    [self updateSections];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    for (BaseAssemblySectionModel *section in self.assemblyComponents.allValues) {
        [section.component assemblyComponentWillAppear:animated];
    }
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    for (BaseAssemblySectionModel *section in self.assemblyComponents.allValues) {
        [section.component assemblyComponentWillDisappear:animated];
    }
}

#pragma mark - work
- (NSArray *)components{
    return @[];
}

#pragma mark - BaseAssemblyTableViewDelagate

- (void)tableViewWillReloadData:(BaseAssemblyTableView *)tableView{
    [self updateSections];
}


/**
 更新sections数据
 */
- (void)updateSections{
    __block BaseAssemblyDispatcher *_wself = self;
    NSArray *list = [self components];
    [self cleanRepeatCompontKey:list];
    [list enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        NSArray *arr = (NSArray *)obj;
        if (!arr && arr.count<1) {
            return ;
        }
        BaseAssemblySectionModel *section = [[BaseAssemblySectionModel alloc] init];
        section.sectionIndexInTableView = idx;
        
        NSString *componentType = arr.firstObject;
        Class componentClass = NSClassFromString(arr[1]);
        BaseAssemblyComponent *component = _wself.assemblyComponents[@(idx).stringValue];
        if (!component || ![component isKindOfClass:componentClass]) {
            component = [self createComponentWithComponentClass:componentClass ComponentType:componentType];
        }
        
        section.component = component;
        section.componentKey = componentType;
        [_wself.assemblyComponents setValue:section forKey:@(idx).stringValue];
    }];
}


/**
 根据组件类创建组件

 @param componentClass 组件类
 @param componentType 组件类型
 @return 组件
 */
- (BaseAssemblyComponent *)createComponentWithComponentClass:(Class)componentClass ComponentType:(NSString *)componentType{
    if ([componentClass isSubclassOfClass:[BaseAssemblyComponent class]]) {
        BaseAssemblyComponent *component = [componentClass new];
        component.tableView = self.tableView;
        component.dataDictionary = self.dataDictionary;
        component.componentType = componentType;
        component.componentSections = [self components];
        [self addChildViewController:component];
        [component setupComponent];
        
        return component;
    }else{
        return [BaseAssemblyComponent new];
    }
}


/**
 清除重复的组件key

 @param list key数组
 */
- (void)cleanRepeatCompontKey:(NSArray *)list{
    __block BaseAssemblyDispatcher *_wself = self;
    [list enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        NSArray *arr = (NSArray *)obj;
        if (!arr && arr.count<1) {
            return ;
        }
        NSString *componentType = arr.firstObject;
        NSArray *values = _wself.assemblyComponents.allValues;
        NSInteger isRepeat = [_wself isRepeat:values withType:componentType];
        if (isRepeat > -1) {
            BaseAssemblySectionModel *section = [self.assemblyComponents objectForKey:@(isRepeat).stringValue];
            [section.component removeFromParentViewController];
            [_wself.assemblyComponents removeObjectForKey:@(isRepeat).stringValue];
        }
    }];
}


/**
 判断是否重复

 @param values 值数组
 @param componentType 组件类型
 @return 重复的值
 */
- (NSInteger)isRepeat:(NSArray *)values withType:(NSString *)componentType{
    __block NSInteger isRepeat = -1;
    [values enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        BaseAssemblySectionModel *section = (BaseAssemblySectionModel *)obj;
        if ([section.componentKey isEqualToString:componentType]) {
            isRepeat = section.sectionIndexInTableView;
            *stop = YES;
        }
    }];
    return isRepeat;
}

#pragma mark - TB

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    NSArray *arr = self.assemblyComponents.allKeys;
    return arr.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    NSString *idx = @(section).stringValue;
    BaseAssemblySectionModel *sectionComponent = [self.assemblyComponents valueForKey:idx];
    return [sectionComponent.component tableView:self.tableView numberOfRowsInSection:section];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    NSString *idx = @(indexPath.section).stringValue;
    BaseAssemblySectionModel *sectionComponent = [self.assemblyComponents valueForKey:idx];
    [sectionComponent.component view];
    return [sectionComponent.component tableView:self.tableView cellForRowAtIndexPath:indexPath];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    NSString *idx = @(indexPath.section).stringValue;
    BaseAssemblySectionModel *sectionComponent = [self.assemblyComponents valueForKey:idx];
    [sectionComponent.component tableView:self.tableView
                  didSelectRowAtIndexPath:indexPath];
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section{
    NSString *idx = @(section).stringValue;
    BaseAssemblySectionModel *sectionComponent = [self.assemblyComponents valueForKey:idx];
    return [sectionComponent.component tableView:self.tableView heightForHeaderInSection:section];
}

- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section{
    NSString *idx = @(section).stringValue;
    BaseAssemblySectionModel *sectionComponent = [self.assemblyComponents valueForKey:idx];
    return [sectionComponent.component tableView:self.tableView heightForFooterInSection:section];
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section{
    NSString *idx = @(section).stringValue;
    BaseAssemblySectionModel *sectionComponent = [self.assemblyComponents valueForKey:idx];
    return [sectionComponent.component tableView:self.tableView viewForHeaderInSection:section];
}

- (UIView *)tableView:(UITableView *)tableView viewForFooterInSection:(NSInteger)section{
    NSString *idx = @(section).stringValue;
    BaseAssemblySectionModel *sectionComponent = [self.assemblyComponents valueForKey:idx];
    return [sectionComponent.component tableView:self.tableView viewForFooterInSection:section];
}

#pragma mark - no used
- (void)tableView:(UITableView *)tableView accessoryButtonTappedForRowWithIndexPath:(NSIndexPath *)indexPath{
    NSString *idx = @(indexPath.section).stringValue;
    BaseAssemblySectionModel *sectionComponent = [self.assemblyComponents valueForKey:idx];
    [sectionComponent.component tableView:self.tableView accessoryButtonTappedForRowWithIndexPath:indexPath];
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath{
    NSString *idx = @(indexPath.section).stringValue;
    BaseAssemblySectionModel *sectionComponent = [self.assemblyComponents valueForKey:idx];
    return [sectionComponent.component tableView:self.tableView canEditRowAtIndexPath:indexPath];
}

- (BOOL)tableView:(UITableView *)tableView canFocusRowAtIndexPath:(NSIndexPath *)indexPath{
    NSString *idx = @(indexPath.section).stringValue;
    BaseAssemblySectionModel *sectionComponent = [self.assemblyComponents valueForKey:idx];
    return [sectionComponent.component tableView:self.tableView canFocusRowAtIndexPath:indexPath];
}

- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath{
    NSString *idx = @(indexPath.section).stringValue;
    BaseAssemblySectionModel *sectionComponent = [self.assemblyComponents valueForKey:idx];
    return [sectionComponent.component tableView:self.tableView canMoveRowAtIndexPath:indexPath];
}
@end

File: TableViewReview/TableViewReview/CompinentTableView/SwiftComponent/SwiftComponent/BaseAssemblyDispatcher.swift
//
//  BaseAssemblyDispatcher.swift
//  TableViewReview
//
//  Created by Jialin Chen on 2019/8/1.
//  Copyright © 2019年 CJL. All rights reserved.
//

import UIKit

class BaseAssemblyDispatcher: UIViewController, BaseAssemblyTableViewDelegate, UITableViewDelegate, UITableViewDataSource {
    
    
    var dataDictionary : Dictionary<String, Any>?
    var tableView : BaseAssemblyTableView?
    
    private var assemblyComponents : Dictionary<String, Any>?
    
    //MARK:-----work
    func components()->NSArray{
        return []
    }

    //MARK:-----Private
    private init() {
        super.init(nibName: nil, bundle: nil)
        self.initWithData()
    }
    
    internal override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        self.initWithData()
    }
    
    internal required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func initWithData(){
        self.assemblyComponents = Dictionary<String, Any>()
    }
    
    internal override func loadView(){
        let rect = UIScreen.main.bounds
        self.view = UIView.init(frame: CGRect(x: 0, y: 0, width: rect.size.width, height: rect.size.height))
        self.view.backgroundColor = UIColor.white
        self.view.autoresizingMask = UIView.AutoresizingMask(rawValue: (UIView.AutoresizingMask.flexibleWidth.rawValue | UIView.AutoresizingMask.flexibleHeight.rawValue))
        self.tableView = BaseAssemblyTableView.init(frame: self.view.bounds, style: .plain)
        self.tableView?.delegate = self
        self.tableView?.delegate = self
        self.tableView?.baseDelegate = self
        self.view.addSubview(self.tableView!)
        
    }
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
       self.initWithData()
    
    }
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        for section in (self.assemblyComponents?.values)! {
            (section as! BaseAssemblySectionModel).component?.assemblyComponentWillAppear(animated)
        }
    }
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        for section in (self.assemblyComponents?.values)! {
            (section as! BaseAssemblySectionModel).component?.assemblyComponentWillDisappear(animated)
        }
    }
    
    //MARK:-----BaseAssemblyTableViewDelegate
    func tableViewWillReloadData(_ tableView: BaseAssemblyTableView) {
        self.updateSections()
    }
    func updateSections(){
        weak var _wself = self
        let list = self.components()
        self.cleanRepeatComponentKey(list)
        list.enumerateObjects { (obj, idx:Int, stop) in
            let arr = obj as! NSArray
            if arr.count < 1 {return}
            let section = BaseAssemblySectionModel()
            section.sectionIndexInTableView = idx
            
            let componentType : String = arr.firstObject as! String
            let componentClass : AnyClass = NSClassFromString(arr[1] as! String)!
            var component : BaseAssemblyComponent? = _wself?.assemblyComponents![String(idx)] as? BaseAssemblyComponent
            if component==nil || (component?.isKind(of: componentClass))!{
                component = self.createComponentWithComponentClass(componentClass, ComponentType: componentType)
            }
            section.component = component
            section.componentKey = componentType
            _wself?.assemblyComponents![String(idx)] = section
        }
    }
    func createComponentWithComponentClass(_ componentClass: AnyClass, ComponentType componentType: String)->BaseAssemblyComponent{
        if componentClass.isSubclass(of: BaseAssemblyComponent.classForCoder()){
            let component : BaseAssemblyComponent = componentClass.init() as! BaseAssemblyComponent
            component.tableView = self.tableView
            component.dataDictionary = self.dataDictionary
            component.componentType = componentType
            component.componentSections = self.components()
            self.addChild(component)
            component.setupComponent()
            
            return component
        }else{
            return BaseAssemblyComponent()
        }
    }
    func cleanRepeatComponentKey(_ list : NSArray){
        weak var _wself = self
        list.enumerateObjects { (obj, idx:Int, stop) in
            let arr = obj as! NSArray
            if arr.count < 1 {return}
            
            let componentType : String = arr.firstObject as! String
            let values : NSArray = NSArray.init(objects: _wself?.assemblyComponents?.values as Any)
            let isRepeate : Int = (_wself?.isRepeat(values, withType: componentType))!
            if isRepeate > -1{
                let section : BaseAssemblySectionModel = self.assemblyComponents![String(idx)] as! BaseAssemblySectionModel
                section.component?.removeFromParent()
                _wself?.assemblyComponents?.removeValue(forKey: String(idx))
            }
        }
    }
    func isRepeat(_ values: NSArray, withType componentType: String)->Int{
        var isRepeate : Int = -1
        values.enumerateObjects { (obj, idx, stop) in
            let section : BaseAssemblySectionModel = obj as! BaseAssemblySectionModel
            if section.componentKey == componentType{
                isRepeate = section.sectionIndexInTableView!
                stop[0] = true
            }
        }
        return isRepeate
    }
    
    //MARK:-----TB
    func numberOfSections(in tableView: UITableView) -> Int {
        let arr = self.assemblyComponents?.keys
        return (arr?.count)!
    }
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let idx = String(section)
        let sectionComponent : BaseAssemblySectionModel = self.assemblyComponents![idx] as! BaseAssemblySectionModel
        return (sectionComponent.component?.tableView(self.tableView!, numberOfRowsInSection: section))!
    }
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let idx = String(indexPath.section)
        let sectionComponent : BaseAssemblySectionModel = self.assemblyComponents![idx] as! BaseAssemblySectionModel
//        sectionComponent.component?.view
        return (sectionComponent.component?.tableView(self.tableView!,cellForRowAt:indexPath))!
    }
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let idx = String(indexPath.section)
        let sectionComponent : BaseAssemblySectionModel = self.assemblyComponents![idx] as! BaseAssemblySectionModel
        sectionComponent.component?.tableView(self.tableView!, didSelectRowAt: indexPath)
    }
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        let idx = String(section)
        let sectionComponent : BaseAssemblySectionModel = self.assemblyComponents![idx] as! BaseAssemblySectionModel
        return (sectionComponent.component?.tableView(self.tableView!, heightForHeaderInSection:section))!
        
    }
    func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        let idx = String(section)
        let sectionComponent : BaseAssemblySectionModel = self.assemblyComponents![idx] as! BaseAssemblySectionModel
        return (sectionComponent.component?.tableView(self.tableView!, heightForFooterInSection:section))!
    }
    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let idx = String(section)
        let sectionComponent : BaseAssemblySectionModel = self.assemblyComponents![idx] as! BaseAssemblySectionModel
        return (sectionComponent.component?.tableView(self.tableView!, viewForFooterInSection:section))!
    }
    func tableView(_ tableView: UITableView, viewForFooterInSection section: Int) -> UIView? {
        let idx = String(section)
        let sectionComponent : BaseAssemblySectionModel = self.assemblyComponents![idx] as! BaseAssemblySectionModel
        return (sectionComponent.component?.tableView(self.tableView!, viewForFooterInSection:section))!
    }
    
    //MARK:-----no used
    func tableView(_ tableView: UITableView, accessoryButtonTappedForRowWith indexPath: IndexPath) {
        let idx = String(indexPath.section)
        let sectionComponent : BaseAssemblySectionModel = self.assemblyComponents![idx] as! BaseAssemblySectionModel
        (sectionComponent.component?.tableView(self.tableView!, accessoryButtonTappedForRowWith:indexPath))!
    }
    func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        let idx = String(indexPath.section)
        let sectionComponent : BaseAssemblySectionModel = self.assemblyComponents![idx] as! BaseAssemblySectionModel
        return (sectionComponent.component?.tableView(self.tableView!, canEditRowAt:indexPath))!
    }
    func tableView(_ tableView: UITableView, canFocusRowAt indexPath: IndexPath) -> Bool {
        let idx = String(indexPath.section)
        let sectionComponent : BaseAssemblySectionModel = self.assemblyComponents![idx] as! BaseAssemblySectionModel
        return (sectionComponent.component?.tableView(self.tableView!, canFocusRowAt:indexPath))!
    }
    func tableView(_ tableView: UITableView, canMoveRowAt indexPath: IndexPath) -> Bool {
        let idx = String(indexPath.section)
        let sectionComponent : BaseAssemblySectionModel = self.assemblyComponents![idx] as! BaseAssemblySectionModel
        return (sectionComponent.component?.tableView(self.tableView!, canMoveRowAt:indexPath))!
    }
}
